nil pointer dereference in federatedseed controller

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|operations|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
"/priority" identifiers: normal|critical|blocker
-->
/kind bug

**What happened**:

```
$ k -n garden logs gardenlet-68976689d5-pn462

E0224 07:56:30.062058       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 6155 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1f39e80, 0x33ac580)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x89
panic(0x1f39e80, 0x33ac580)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions.getShootStateExtensionStateAndResources(0x0, 0x2201476, 0xb, 0xc02e5e9450, 0x0, 0xc003048ac0, 0x3e, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go:178 +0x26
github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions.(*ShootStateControl).CreateShootStateSyncReconcileFunc.func1(0x24e0ec0, 0xc02e6b27c0, 0x0, 0x0, 0xc003048ac0, 0x3e, 0x0, 0x0, 0xc02dcece30, 0x0)
	/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go:91 +0x350
sigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile(0xc02ff60270, 0x24e0ec0, 0xc02e6b27c0, 0x0, 0x0, 0xc003048ac0, 0x3e, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/reconcile/reconcile.go:102 +0x6f
github.com/gardener/gardener/pkg/controllerutils.worker.func1(0x24fe260, 0xc02e998880, 0x2487820, 0xc02ff60270, 0x24e0ec0, 0xc02e6b27c0, 0xc02ff5a220, 0x16, 0x439400)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:114 +0x13e
github.com/gardener/gardener/pkg/controllerutils.worker(0x24e0ec0, 0xc02e6b27c0, 0x24fe260, 0xc02e998880, 0xc02ff5a220, 0x16, 0x2487820, 0xc02ff60270)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:131 +0x7a
github.com/gardener/gardener/pkg/controllerutils.CreateWorker.func2.1(0x24e0ec0, 0xc02e6b27c0)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:72 +0x6b
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc001c9c6e8)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00a8e9ee8, 0x2485200, 0xc02e5f94d0, 0xc02e6b2701, 0xc02e53d0e0)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001c9c6e8, 0x3b9aca00, 0x0, 0xc01bf25b01, 0xc02e53d0e0)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x24e0ec0, 0xc02e6b27c0, 0xc001c9c748, 0x3b9aca00, 0x0, 0x1)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(...)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:99
github.com/gardener/gardener/pkg/controllerutils.CreateWorker.func2(0x24e0ec0, 0xc02e6b27c0, 0x24fe260, 0xc02e998880, 0xc02ff5a220, 0x16, 0x2487820, 0xc02ff60270, 0xc02e772fc0, 0xc02e72d508)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:71 +0xd2
created by github.com/gardener/gardener/pkg/controllerutils.CreateWorker
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:70 +0x285
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x130 pc=0x1cc7de6]

goroutine 6155 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x10c
panic(0x1f39e80, 0x33ac580)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions.getShootStateExtensionStateAndResources(0x0, 0x2201476, 0xb, 0xc02e5e9450, 0x0, 0xc003048ac0, 0x3e, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go:178 +0x26
github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions.(*ShootStateControl).CreateShootStateSyncReconcileFunc.func1(0x24e0ec0, 0xc02e6b27c0, 0x0, 0x0, 0xc003048ac0, 0x3e, 0x0, 0x0, 0xc02dcece30, 0x0)
	/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go:91 +0x350
sigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile(0xc02ff60270, 0x24e0ec0, 0xc02e6b27c0, 0x0, 0x0, 0xc003048ac0, 0x3e, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/reconcile/reconcile.go:102 +0x6f
github.com/gardener/gardener/pkg/controllerutils.worker.func1(0x24fe260, 0xc02e998880, 0x2487820, 0xc02ff60270, 0x24e0ec0, 0xc02e6b27c0, 0xc02ff5a220, 0x16, 0x439400)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:114 +0x13e
github.com/gardener/gardener/pkg/controllerutils.worker(0x24e0ec0, 0xc02e6b27c0, 0x24fe260, 0xc02e998880, 0xc02ff5a220, 0x16, 0x2487820, 0xc02ff60270)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:131 +0x7a
github.com/gardener/gardener/pkg/controllerutils.CreateWorker.func2.1(0x24e0ec0, 0xc02e6b27c0)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:72 +0x6b
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc001c9c6e8)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00a8e9ee8, 0x2485200, 0xc02e5f94d0, 0xc02e6b2701, 0xc02e53d0e0)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001c9c6e8, 0x3b9aca00, 0x0, 0xc01bf25b01, 0xc02e53d0e0)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x24e0ec0, 0xc02e6b27c0, 0xc001c9c748, 0x3b9aca00, 0x0, 0x1)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(...)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:99
github.com/gardener/gardener/pkg/controllerutils.CreateWorker.func2(0x24e0ec0, 0xc02e6b27c0, 0x24fe260, 0xc02e998880, 0xc02ff5a220, 0x16, 0x2487820, 0xc02ff60270, 0xc02e772fc0, 0xc02e72d508)
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:71 +0xd2
created by github.com/gardener/gardener/pkg/controllerutils.CreateWorker
	/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:70 +0x285
```

It seems to happen after https://github.com/gardener/gardener/pull/3550

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:

- Gardener version: master
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

--- a/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go
+++ b/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go
@@ -82,7 +82,15 @@ func (s *ShootStateControl) CreateShootStateSyncReconcileFunc(kind string, objec
 		newState := extensionObj.GetExtensionStatus().GetState()
 		newResources := extensionObj.GetExtensionStatus().GetResources()
 
-		shootState, err := s.getShootState(ctx, req)
+		cluster, err := s.getClusterFromRequest(ctx, req)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return reconcile.Result{}, nil
+			}
+			return reconcile.Result{}, fmt.Errorf("could not get cluster with name %s : %v", cluster.Name, err)
+		}
+
+		shootState, err := s.getShootStateFromCluster(ctx, cluster)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
@@ -227,7 +235,7 @@ func (s *ShootStateControl) getExtensionObject(ctx context.Context, key types.Na
 	return apiextensions.Accessor(obj)
 }
 
-func (s *ShootStateControl) getShootState(ctx context.Context, req reconcile.Request) (*gardencorev1alpha1.ShootState, error) {
+func (s *ShootStateControl) getClusterFromRequest(ctx context.Context, req reconcile.Request) (*extensionsv1alpha1.Cluster, error) {
 	var clusterName string
 	if req.Namespace == "" {
 		// Handling for cluster-scoped backupentry extension resources.
@@ -238,12 +246,13 @@ func (s *ShootStateControl) getShootState(ctx context.Context, req reconcile.Req
 
 	cluster := &extensionsv1alpha1.Cluster{}
 	if err := s.seedClient.Client().Get(ctx, kutil.Key(clusterName), cluster); err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil, nil
-		}
-		return nil, fmt.Errorf("could not get cluster with name %s : %v", clusterName, err)
+		return nil, err
 	}
 
+	return cluster, nil
+}
+
+func (s *ShootStateControl) getShootStateFromCluster(ctx context.Context, cluster *extensionsv1alpha1.Cluster) (*gardencorev1alpha1.ShootState, error) {
 	shoot, err := s.shootRetriever.FromCluster(cluster)
 	if err != nil {
 		return nil, err

--- a/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control_test.go
+++ b/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control_test.go
@@ -173,14 +173,13 @@ var _ = Describe("ShootState Control", func() {
 				},
 			},
 		}
-
-		Expect(seedClient.Create(ctx, cluster)).To(Succeed())
 	})
 
 	Describe("#CreateShootStateSyncReconcileFunc", func() {
 		It("should properly update ShootState with extension state", func() {
 			Expect(fakeGardenClient.Client().Create(ctx, shootState)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, extension)).To(Succeed())
+			Expect(fakeSeedClient.Client().Create(ctx, cluster)).To(Succeed())
 
 			reconcilerFunc := shootStateControl.CreateShootStateSyncReconcileFunc(extensionsv1alpha1.ExtensionResource, extensionObjCreator)
 			_, err := reconcilerFunc.Reconcile(ctx, reconcileRequest)
@@ -198,6 +197,7 @@ var _ = Describe("ShootState Control", func() {
 			Expect(fakeGardenClient.Client().Create(ctx, shootState)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, extension)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, secretObj)).To(Succeed())
+			Expect(fakeSeedClient.Client().Create(ctx, cluster)).To(Succeed())
 
 			reconcilerFunc := shootStateControl.CreateShootStateSyncReconcileFunc(extensionsv1alpha1.ExtensionResource, extensionObjCreator)
 			_, err := reconcilerFunc.Reconcile(ctx, reconcileRequest)
@@ -214,6 +214,7 @@ var _ = Describe("ShootState Control", func() {
 
 			Expect(fakeGardenClient.Client().Create(ctx, shootStateWithExtensionData)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, extension)).To(Succeed())
+			Expect(fakeSeedClient.Client().Create(ctx, cluster)).To(Succeed())
 
 			reconcilerFunc := shootStateControl.CreateShootStateSyncReconcileFunc(extensionsv1alpha1.ExtensionResource, extensionObjCreator)
 			_, err := reconcilerFunc.Reconcile(ctx, reconcileRequest)
@@ -237,6 +238,7 @@ var _ = Describe("ShootState Control", func() {
 			Expect(fakeGardenClient.Client().Create(ctx, shootStateWithExtensionData)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, extension)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, newSecretObj)).To(Succeed())
+			Expect(fakeSeedClient.Client().Create(ctx, cluster)).To(Succeed())
 
 			reconcilerFunc := shootStateControl.CreateShootStateSyncReconcileFunc(extensionsv1alpha1.ExtensionResource, extensionObjCreator)
 			_, err := reconcilerFunc.Reconcile(ctx, reconcileRequest)
@@ -251,6 +253,7 @@ var _ = Describe("ShootState Control", func() {
 			extension.Status.State = nil
 			Expect(fakeGardenClient.Client().Create(ctx, shootStateWithExtensionData)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, extension)).To(Succeed())
+			Expect(fakeSeedClient.Client().Create(ctx, cluster)).To(Succeed())
 
 			reconcilerFunc := shootStateControl.CreateShootStateSyncReconcileFunc(extensionsv1alpha1.ExtensionResource, extensionObjCreator)
 			_, err := reconcilerFunc.Reconcile(ctx, reconcileRequest)
@@ -268,6 +271,7 @@ var _ = Describe("ShootState Control", func() {
 
 			Expect(fakeGardenClient.Client().Create(ctx, shootStateWithExtensionData)).To(Succeed())
 			Expect(fakeSeedClient.Client().Create(ctx, extension)).To(Succeed())
+			Expect(fakeSeedClient.Client().Create(ctx, cluster)).To(Succeed())
 
 			reconcilerFunc := shootStateControl.CreateShootStateSyncReconcileFunc(extensionsv1alpha1.ExtensionResource, extensionObjCreator)
 			_, err := reconcilerFunc.Reconcile(ctx, reconcileRequest)
@@ -294,6 +298,14 @@ var _ = Describe("ShootState Control", func() {
 			Expect(err).NotTo(HaveOccurred())
 
 			ctrl.Finish()
+		})
+
+		It("should not throw any errors if Cluster resource does not exists", func() {
+			Expect(fakeSeedClient.Client().Create(ctx, extension)).To(Succeed())
+
+			reconcilerFunc := shootStateControl.CreateShootStateSyncReconcileFunc(extensionsv1alpha1.ExtensionResource, extensionObjCreator)
+			_, err := reconcilerFunc.Reconcile(ctx, reconcileRequest)
+			Expect(err).NotTo(HaveOccurred())
 		})
 	})
 })

